Support for `VecMonitor` for gym3-style environments

vwxyzjn opened the pull request:
## Description

Gym3-style environments such a procgen directly produces the vectorized environments, so there is no chance of passing a Monitor wrapper to the envs creation process. However, there still is a need to record the episodic stats. One potential way to do this is through the VecMonitor wrapper as done in openai/baselines. See #302 for further related discussions.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- You can use the syntax `closes #100` if this solves the issue #100 -->
- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Was having some troubles with the  `make type`. It seems the error was not caused by my changes, maybe?

```
[50/91] check stable_baselines3.common.base_class
FAILED: /home/costa/Documents/work/go/src/github.com/vwxyzjn/stable-baselines3/.pytype/pyi/stable_baselines3/common/base_class.pyi 
/home/costa/anaconda3/bin/python -m pytype.single --imports_info /home/costa/Documents/work/go/src/github.com/vwxyzjn/stable-baselines3/.pytype/imports/stable_baselines3.common.base_class.imports --module-name stable_baselines3.common.base_class -V 3.8 -o /home/costa/Documents/work/go/src/github.com/vwxyzjn/stable-baselines3/.pytype/pyi/stable_baselines3/common/base_class.pyi --analyze-annotated --nofail --quick /home/costa/Documents/work/go/src/github.com/vwxyzjn/stable-baselines3/stable_baselines3/common/base_class.py
File "/home/costa/Documents/work/go/src/github.com/vwxyzjn/stable-baselines3/stable_baselines3/common/base_class.py", line 627, in load: Can't instantiate BaseAlgorithm with abstract methods _setup_model, learn [not-instantiable]
File "/home/costa/Documents/work/go/src/github.com/vwxyzjn/stable-baselines3/stable_baselines3/common/base_class.py", line 627, in load: Invalid keyword argument _init_setup_model to function BaseAlgorithm.__init__ [wrong-keyword-args]
         Expected: (self, policy, env, policy_base, learning_rate, policy_kwargs, tensorboard_log, verbose, device, support_multi_env, create_eval_env, monitor_wrapper, seed, use_sde, sde_sample_freq, supported_action_spaces)
  Actually passed: (self, policy, env, device, _init_setup_model)

For more details, see https://google.github.io/pytype/errors.html.
ninja: build stopped: subcommand failed.
Leaving directory '/home/costa/Documents/work/go/src/github.com/vwxyzjn/stable-baselines3/.pytype'
make: *** [Makefile:8: type] Error 1
```


Given this new feature, the following two ways of initializing vectorized envs should be equivalent, except the latter is *much faster* when used with higher number of parallel envs.

```python
from stable_baselines3.common.monitor import Monitor
from stable_baselines3.common.vec_env import DummyVecEnv
import gym
def make_env(gym_id):
    def thunk():
        env = gym.make(gym_id, render_mode='rgb_array')
        env = Monitor(env)
        return env
    return thunk
envs = DummyVecEnv([make_env('procgen-starpilot-v0')])
```

```python
from procgen import ProcgenEnv
from stable_baselines3.common.vec_env import VecExtractDictObs, VecMonitor
venv = ProcgenEnv(num_envs=1, env_name='starpilot')
venv = VecExtractDictObs(venv, "rgb")
venv = VecMonitor(venv=venv)
```

vwxyzjn commented:
Just saw the CI failed, however the doc build successfully on my end. Hmmm.

```
(base) ~/.../vwxyzjn/stable-baselines3 >>> cd docs                                                                                                                                             
(base) ~/.../stable-baselines3/docs >>> make html                                                                                                                                              
Running Sphinx v3.1.2

** (process:21786): WARNING **: 19:33:00.337: Error loading plugin: libhspell.so.0: cannot open shared object file: No such file or directory


** (process:21786): WARNING **: 19:33:00.337: Error loading plugin: libnuspell.so.4: cannot open shared object file: No such file or directory


** (process:21786): WARNING **: 19:33:00.337: Error loading plugin: libvoikko.so.1: cannot open shared object file: No such file or directory


** (process:21786): WARNING **: 19:33:00.337: Error loading plugin: libaspell.so.15: cannot open shared object file: No such file or directory

Initializing Spelling Checker 7.1.0
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: 0 added, 0 changed, 0 removed
looking for now-outdated files... none found
no targets are out of date.
build succeeded.

The HTML pages are in _build/html.
```

araffin commented:
The error for the doc:

```
/home/runner/work/stable-baselines3/stable-baselines3/stable_baselines3/common/vec_env/vec_monitor.py:docstring of stable_baselines3.common.vec_env.vec_monitor.VecMonitor:14:Unexpected indentation.

```

make sure to have the latest package version of sphinx ;)

araffin commented on stable_baselines3/common/vec_env/vec_extract_dict_obs.py:
```diff
@@ -0,0 +1,26 @@
+
+class VecExtractDictObs(VecEnvWrapper):
+    """
+    A vectorized monitor wrapper for extracting dictionary observations.
```
Why is it a monitor wrapper?

vwxyzjn replied:
Typo. Fixed

araffin commented on stable_baselines3/common/vec_env/vec_monitor.py:
```diff
@@ -0,0 +1,72 @@
+    """
+    A vectorized monitor wrapper for *vectorized* Gym environments, it is used to record the episode reward, length, time and other data.
+
+    Some environments like [`openai/procgen`](https://github.com/openai/procgen) or [`gym3`](https://github.com/openai/gym3) directly
```
you should use rst format here, not markdown ;)

vwxyzjn replied:
Nice catch. Fixed

araffin commented on stable_baselines3/common/vec_env/vec_monitor.py:
```diff
@@ -0,0 +1,72 @@
+
+    def __init__(self, venv: VecEnv):
+        VecEnvWrapper.__init__(self, venv)
+        self.eprets = None
```
Please use more informative names like `episode_returns`, `episode_lengths`, `start_time` ;)

vwxyzjn replied:
I have used `episode_returns, episode_lengths, episode_count, t_start`. `t_start` is consistent with the usage here.

https://github.com/DLR-RM/stable-baselines3/blob/b01bde3e2d6f6304a0ec92b0bd00564951e753ce/stable_baselines3/common/monitor.py#L40

araffin commented on stable_baselines3/common/vec_env/vec_monitor.py:
```diff
@@ -0,0 +1,72 @@
+
+    def reset(self) -> VecEnvObs:
+        obs = self.venv.reset()
+        self.eprets = np.zeros(self.num_envs, "f")
```
please use `dtype=np.float32` (more explicit than "f")

vwxyzjn replied:
fixed

araffin commented on stable_baselines3/common/vec_env/vec_monitor.py:
```diff
@@ -0,0 +1,72 @@
+        return obs
+
+    def step_wait(self) -> VecEnvStepReturn:
+        obs, rews, dones, infos = self.venv.step_wait()
```
please use `rewards`

vwxyzjn replied:
fixed

araffin commented on stable_baselines3/common/vec_env/vec_monitor.py:
```diff
@@ -0,0 +1,72 @@
+                ret = self.eprets[i]
+                eplen = self.eplens[i]
+                epinfo = {"r": ret, "l": eplen, "t": round(time.time() - self.tstart, 6)}
+                info["episode"] = epinfo
```
you should probably check at the very beginning if a monitor wrapper is already present or not (cf what we do with `evaluate_policy`) otherwise it would overwrite it.

vwxyzjn replied:
I am not sure how to do the check exactly...

araffin replied:
you may take a look at https://github.com/DLR-RM/stable-baselines3/blob/master/tests/test_vec_envs.py#L432
and https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/evaluation.py#L60

araffin reviewed (changes requested):
Overall look good, please add some tests and the csv logging feature ;)

Note: we are using rst for the dosctring format, not markdown unfortunately

vwxyzjn commented:
@araffin Thank you for the very helpful review. I have addressed most concerns raised in your comments. Please take a look when you have a chance :)

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -38,33 +41,26 @@ def __init__(
-        if filename is None:
-            self.file_handler = None
-            self.logger = None
+        if filename:
```
```suggestion
        if filename is not None:
```

araffin commented on docs/guide/vec_envs.rst:
```diff
@@ -90,3 +90,15 @@ VecTransposeImage
 
 .. autoclass:: VecTransposeImage
   :members:
+
```
as a general remark, as mentioned in the PR template and contributing guide, please update the changelog accordingly too

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -38,33 +41,26 @@ def __init__(
-            self.file_handler = None
-            self.logger = None
+        if filename:
+            self.results_writer = ResultsWriter(
```
Did you check that this method was equivalent to what was there before?

vwxyzjn replied:
Yes, and all the test cases went through.

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -38,33 +41,26 @@ def __init__(
+        if filename:
+            self.results_writer = ResultsWriter(
+                filename,
+                header={"t_start": self.t_start, "env_id": env.spec and env.spec.id},
```
why `env.spec and env.spec.id` ?

vwxyzjn replied:
This is largely taken from `openai/baselines`. I am not very familiar with the CSV related workflow.

araffin replied:
now I get it, only try to access `env.spec.id`, if `env.spec` is not None, otherwise it would throw an error

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -163,6 +158,42 @@ class LoadMonitorResultsError(Exception):
+    :param filename: the location to save a log file, can be None for no log
+    :param header: the header dictionary object of the saved csv
+    :param reset_keywords: the extra information to log, typically is composed of
+        `reset_keywords` and `info_keywords`
```
```suggestion
        ``reset_keywords`` and ``info_keywords``
```

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -163,6 +158,42 @@ class LoadMonitorResultsError(Exception):
+    def __init__(
+        self,
+        filename: str = "",
+        header: Dict[str, Union[float, str]] = {},
```
please avoid mutable as default, see https://github.com/satwikkansal/wtfpython#-beware-of-default-mutable-arguments

vwxyzjn replied:
Should I use `None` instead maybe?

araffin replied:
yes, the classic way of doing it is:

```python
if header is None:
    header = {}
```

araffin commented on tests/test_vec_extract_dict_obs.py:
```diff
@@ -0,0 +1,37 @@
+    """Test VecExtractDictObs"""
+
+    env = DictObsVecEnv()
+    env = VecExtractDictObs(env, "rgb")
```
maybe add a quick run test with an algorithm like A2C/PPO?

araffin replied:
dito for the VecMonitor

vwxyzjn commented:
Hey @araffin sorry for the delay, got stuck in hw and projects, procrastinating like crazy. Anyway I have addressed most of the concerns. I also had a chance to try running the procgen env, but was having a minor issue:

![image](https://user-images.githubusercontent.com/5555347/109575838-98280080-7ac0-11eb-9080-bcab0e43827f.png)

So basically the gym3-style environments don't necessarily return `terminal_observation`, which breaks the `VecTransposeImage` wrapper. Maybe we should treat this as a separate issue from this PR? 

Thanks for the helpful comments. Let me know if you need anything else.

Miffyli commented:
> So basically the gym3-style environments don't necessarily return terminal_observation, which breaks the VecTransposeImage wrapper. Maybe we should treat this as a separate issue from this PR?

Ugh... I assume this would require modifying environments to provide it (as they do the "vectorization" part)? In any case, this should be handled in a separate issue to avoid bloating this one up :)

araffin commented:
>So basically the gym3-style environments don't necessarily return terminal_observation,

hmm, do gym3 style environments reset by themselves? or how do they handle terminal observations then?

>? In any case, this should be handled in a separate issue to avoid bloating this one up :)

agree with that ;)

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -1,11 +1,14 @@
-__all__ = ["Monitor", "get_monitor_files", "load_results"]
+# flake8: noqa F401
```
why for the whole file? (unused import?)

vwxyzjn replied:
I am not sure... I noticed `get_monitor_files` and others are in `__all__` so I just included everything similarly like

```__all__ = ["Monitor", "ResultsWriter", "get_monitor_files", "load_results"]```

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -85,7 +81,7 @@ def reset(self, **kwargs) -> GymObs:
         return self.env.reset(**kwargs)
 
-    def step(self, action: Union[np.ndarray, int]) -> GymStepReturn:
+    def step(self, action: Union[np.ndarray, int]) -> "GymStepReturn":
```
why as string? what was causing an error?

vwxyzjn replied:
So when I import the `ResultsWriter` in the `vec_monitor.py`, it causes a circular import.

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -163,6 +158,42 @@ class LoadMonitorResultsError(Exception):
+        self.logger.writeheader()
+        self.file_handler.flush()
+
+    def write_row(self, epinfo):
```
missing docstring + type hints

vwxyzjn replied:
fixed

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -163,6 +158,42 @@ class LoadMonitorResultsError(Exception):
+            self.logger.writerow(epinfo)
+            self.file_handler.flush()
+
+    def close(self):
```
missing type hint

vwxyzjn replied:
fixed

araffin commented on stable_baselines3/common/vec_env/vec_monitor.py:
```diff
@@ -0,0 +1,74 @@
+        self.episode_count = 0
+        self.t_start = time.time()
+        if filename:
+            self.results_writer = ResultsWriter(filename, header={"t_start": self.t_start}, extra_keys=info_keywords)
```
no env id that time ?

vwxyzjn replied:
![image](https://user-images.githubusercontent.com/5555347/110230263-b8e7c000-7edd-11eb-9053-dbf7d49583ce.png)

So gym3 style envs don't necessarily have the field.

araffin replied:
I see, but you can still do a check and use it if present, to be consistent with `Monitor`

vwxyzjn commented:
> hmm, do gym3 style environments reset by themselves? or how do they handle terminal observations then?

The answer is [here](https://github.com/openai/gym3/blob/master/docs/design.md)

- **No reset function: the environment resets itself**.
    - In `gym.Env`, running the environment after `done=True` but without resetting led to undefined behavior.
    - Another annoyance of `gym.Env` and `VecEnv` was that given an environment, there was no way to get the first observation, which the first action would depend on, without resetting the environment.
    - It's perfectly valid to define an environment without a reset function, and general RL algorithms don't need to use a reset function. Hence, it makes sense to leave it out of the required interface. You're welcome to define a reset function, possibly with arguments, if you plan to pair your environment with an algorithm that uses this feature.
    - *What if I want to reset a whole wrapper stack, which contains stateful wrappers like FrameSkip?* You can use a constructor function to rebuild the object. If the construction process is expensive, you can make your constructor return a fixed object.

Also I think by default gym3 does not store the terminal observation. See [here](https://github.com/openai/gym3/blob/4c3824680eaf9dd04dce224ee3d4856429878226/gym3/interop.py#L163-L171)

araffin commented:
>The answer is here

this I knew, but it is still a bit weird to me in the example they show: they act without having any observation... (but I guess you can call `observe()` before)

>Also I think by default gym3 does not store the terminal observation. See here

then I think there might be a bug in the gym3 design... (no terminal observation)
probably better to do a PR there, no?

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -163,6 +158,52 @@ class LoadMonitorResultsError(Exception):
+        self.logger.writeheader()
+        self.file_handler.flush()
+
+    def write_row(self, epinfo: Dict[str, Union[float, int]]):
```
```suggestion
    def write_row(self, epinfo: Dict[str, Union[float, int]]) -> None:
```

araffin commented on stable_baselines3/common/monitor.py:
```diff
@@ -163,6 +158,52 @@ class LoadMonitorResultsError(Exception):
+            self.logger.writerow(epinfo)
+            self.file_handler.flush()
+
+    def close(self):
```
```suggestion
    def close(self) -> None:
```

vwxyzjn commented:
> this I knew, but it is still a bit weird to me in the example they show: they act without having any observation... (but I guess you can call observe() before)

Yeah, I think this was intended for a more async operation.

> then I think there might be a bug in the gym3 design... (no terminal observation)
probably better to do a PR there, no?

Yes, I would agree.

vwxyzjn commented:
Hey, hope you all are doing well. Just checking is there anything else you would like me to add for this PR.

araffin commented:
>Hey, hope you all are doing well. Just checking is there anything else you would like me to add for this PR.

yep, I think there are still some comments not addressed...

vwxyzjn commented:
> apparently, there is one left and then I may give a try to simplify the imports

Sounds good :) Let me know if I can provide any help.

weepingwillowben commented:
I just got another feature request for this PR here: https://github.com/PettingZoo-Team/PettingZoo/issues/363. 

The background is that PettingZoo has support for stable baselines by converting a pettingzoo multiagent environment into an equivalent vector environment https://github.com/PettingZoo-Team/SuperSuit#parallel-environment-vectorization. So this PR will enable reward logging for multiagent environments with SB3, which is clearly very valuable.

vwxyzjn commented:
Just incorporated the fix from #379

vwxyzjn commented:
>  add a warning in the VecEnv doc

I was gonna add a warning in the VecEnv doc but wasn't sure how to add it. Especially because the gym3 vec envs is kind of a separate entities from `VecEnv`...

araffin commented:
I've updated the PR and wanted to push the commits on my side, but I need write access to https://github.com/vwxyzjn/stable-baselines3 ...

araffin reviewed (approved):
LGTM, thanks =)

vwxyzjn commented:
Thanks!
